nibabel/streamlines/tractogram.py

```diff
@@ -0,0 +1,677 @@
+
+    @data_per_streamline.setter
+    def data_per_streamline(self, value):
+        self._data_per_streamline = LazyDict(self, value)
+
+    @property
+    def data_per_point(self):
+        return self._data_per_point
+
+    @data_per_point.setter
+    def data_per_point(self, value):
+        self._data_per_point = LazyDict(self, value)
+
+    @property
+    def data(self):
+        if self._data is not None:
+            return self._data()
+
+        def _gen_data():
+            data_per_streamline_generators = {}
+            for k, v in self.data_per_streamline.items():
+                data_per_streamline_generators[k] = iter(v)
+
+            data_per_point_generators = {}
+            for k, v in self.data_per_point.items():
+                data_per_point_generators[k] = iter(v)
+
+            for s in self.streamlines:
+                data_for_streamline = {}
+                for k, v in data_per_streamline_generators.items():
+                    data_for_streamline[k] = next(v)
+
+                data_for_points = {}
+                for k, v in data_per_point_generators.items():
+                    data_for_points[k] = next(v)
+
+                yield TractogramItem(s, data_for_streamline, data_for_points)
+
+        return _gen_data()
+
+    def __getitem__(self, idx):
+        raise NotImplementedError('`LazyTractogram` does not support indexing.')
+
+    def __iter__(self):
+        i = 0
+        for i, tractogram_item in enumerate(self.data, start=1):
+            yield tractogram_item
+
+        # Keep how many streamlines there are in this tractogram.
+        self._nb_streamlines = i
+
+    def __len__(self):
+        # Check if we know how many streamlines there are.
+        if self._nb_streamlines is None:
+            warn("Number of streamlines will be determined manually by looping"
+                 " through the streamlines. If you know the actual number of"
+                 " streamlines, you might want to set it beforehand via"
+                 " `self.header.nb_streamlines`."
+                 " Note this will consume any generators used to create this"
+                 " `LazyTractogram` object.", Warning)
+            # Count the number of streamlines.
+            self._nb_streamlines = sum(1 for _ in self.streamlines)
+
+        return self._nb_streamlines
+
+    def copy(self):
+        """ Returns a copy of this :class:`LazyTractogram` object. """
+        tractogram = LazyTractogram(self._streamlines,
+                                    self._data_per_streamline,
+                                    self._data_per_point)
+        tractogram._nb_streamlines = self._nb_streamlines
+        tractogram._data = self._data
+        tractogram._affine_to_apply = self._affine_to_apply.copy()
+        return tractogram
+
+    def apply_affine(self, affine):
+        """ Applies an affine transformation to the streamlines.
+
+        The transformation will be applied just before returning the
```

Comment:
I rephrased to: "The transformation given by the `affine` matrix is applied after any pending transformations to the streamline points."